modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -3112,43 +3116,64 @@ void GDScriptAnalyzer::reduce_identifier_from_base(GDScriptParser::IdentifierNod
 	}
 
 	GDScriptParser::ClassNode *base_class = base.class_type;
+	List<GDScriptParser::ClassNode *> script_classes;
+	bool is_base = false;
 
-	// TODO: Switch current class/function/suite here to avoid misrepresenting identifiers (in recursive reduce calls).
-	while (base_class != nullptr) {
-		if (base_class->identifier && base_class->identifier->name == name) {
-			p_identifier->set_datatype(base_class->get_datatype());
+	if (base_class != nullptr) {
+		is_base = true;
+		get_class_node_current_scope_classes(base_class, &script_classes);
+	}
+
+	for (GDScriptParser::ClassNode *script_class : script_classes) {
+		if (script_class->identifier && script_class->identifier->name == name) {
+			p_identifier->set_datatype(script_class->get_datatype());
 			return;
 		}
 
-		if (base_class->has_member(name)) {
-			resolve_class_member(base_class, name, p_identifier);
+		if (script_class->has_member(name)) {
+			resolve_class_member(script_class, name, p_identifier);
 
-			GDScriptParser::ClassNode::Member member = base_class->get_member(name);
+			GDScriptParser::ClassNode::Member member = script_class->get_member(name);
 			p_identifier->set_datatype(member.get_datatype());
```

Comment:
This needs to be moved to individual cases, because the member might be of irrelevant type (like an instance variable of an outer class).